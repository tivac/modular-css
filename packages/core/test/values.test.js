"use strict";

var path = require("path"),
    dedent = require("dedent"),
    namer  = require("test-utils/namer.js"),
    
    Processor = require("../processor.js");

describe("/processor.js", () => {
    describe("values", () => {
        var processor;
        
        beforeEach(() => {
            processor = new Processor({
                namer
            });
        });
        
        it("should fail on invalid value syntax", () =>
            processor.string(
                "./invalid/value.css",
                "@value foo, bar from nowhere.css"
            )
            .catch((error) =>
                expect(error.message).toMatch(`SyntaxError: Expected source but "n" found.`)
            )
        );

        it("should fail if a value imports a non-existant reference", () =>
            processor.string(
                "./invalid/value.css",
                "@value not-real from \"../local.css\";"
            )
            .catch((error) =>
                expect(error.message).toMatch(
                    `Unable to locate "../local.css" from "${path.resolve("invalid/value.css")}"`
                )
            )
        );

        it("should support simple values", () =>
            processor.string(
                "./values.css",
                dedent(`
                    @value a: red;
                    @value b:
                        Segoe UI
                        sans-serif;

                    .a {
                        color: a;
                        font-family: b;
                    }
                `)
            )
            .then(() => processor.output())
            .then((result) =>
                expect(result.css).toMatchSnapshot()
            )
        );

        it("should support local values in value composition", () =>
            processor.string(
                "./packages/core/test/specimens/simple.css",
                dedent(`
                    @value o: one;
                    @value local: './local.css';
                    @value o from local;
                    .fooga { background: one; }
                `)
            )
            .then((result) =>
                expect(result.exports).toMatchSnapshot()
            )
        );

        it("should support value composition", () =>
            processor.file(require.resolve("./specimens/value-composition.css"))
            .then(() => processor.output())
            .then((result) =>
                expect(result.css).toMatchSnapshot()
            )
        );

        it("should support value namespaces", () =>
            processor.file(require.resolve("./specimens/value-namespace.css"))
            .then(() => processor.output())
            .then((result) =>
                expect(result.css).toMatchSnapshot()
            )
        );

        it("should support value replacement in :external(...)", () =>
            processor.file(require.resolve("./specimens/externals.css"))
            .then(() => processor.output())
            .then((result) =>
                expect(result.css).toMatchSnapshot()
            )
        );
    });
});
