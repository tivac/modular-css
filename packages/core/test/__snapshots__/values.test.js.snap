// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/processor.js values should support local values in value composition 1`] = `
Object {
  "fooga": Array [
    "fooga",
  ],
}
`;

exports[`/processor.js values should support simple values 1`] = `
"/* values.css */
.a { color: red; }"
`;

exports[`/processor.js values should support value composition 1`] = `
"/* packages/core/test/specimens/values.css */

/* packages/core/test/specimens/value-composition.css */
.red {
    color: red;
}
"
`;

exports[`/processor.js values should support value namespaces 1`] = `
"/* packages/core/test/specimens/values.css */

/* packages/core/test/specimens/value-namespace.css */
.red {
    color: red;
}
.blue {
    color: blue;
}
"
`;

exports[`/processor.js values should support value replacement in :external(...) 1`] = `
"/* packages/core/test/specimens/folder/folder.css */
.folder { margin: 2px; }

/* packages/core/test/specimens/simple.css */
.wooga { color: red; }

/* packages/core/test/specimens/local.css */
.booga { background: green; }

/* packages/core/test/specimens/externals.css */
.a .wooga {
    color: green;
}
.b:hover .booga {
    background: blue;
}
"
`;
